ARCHITECTURE.md

```diff
@@ -0,0 +1,255 @@
+either of these things.
+
+This kustomization is the access node to a directed, acyclic graph of
+Kubernetes objects, including other kustomizations, to include in a
+build.
+
+Execution of `build` starts and ends in the [`api`] module,
+frequently dipping into the [`kyaml`] module for lower level
+YAML manipulation.
+
+### The `build` flow
+
+  - Validate command lines arguments and flags.
+
+  - Make a `Kustomizer` as a function of those arguments.
+
+  - Call `Run` on the kustomizer, passing the path to the kustomization.
+
+    `Run` returns an instance of `ResMap`, which is the `api`
+    package's representation of a list of kubernetes objects.
+    These objects are fully hydrated, per the instructions in the
+    kustomization.
+
+  - Emit the YAML version of those objects, to files or `stdout`.
+
+
+### The `Run` function
+
+  - Create various objects
+
+    - A `ResMap` factory.
+
+      Makes  `ResMaps` from byte streams, other `ResMaps`, etc.
+
+    - A file `loader.Loader`.
+
+      It's fed an appropriate set of restrictions, and the path to the kustomization.
+
+    - A plugin loader.
+
+      It finds plugins (transformers, generators or validators)
+      and prepare them for running.
+
+    - A `KustTarget` encapsulating all of the above.
+
+      A KustTarget contains one `Kustomization` and represents
+	  everything that kustomization can reach.  This will include
+	  other `KustTarget` instances, each having a smaller purvue than
+	  the one referencing it.
+
+  - Call `KustTarget.Load` to load its kustomization.
+
+    This step deals with deprecations and field changes.
+
+  - Load [openapi] data specified by the kustomization.
+
+    This is needed to recognize k8s kinds and their special
+    properties, e.g. which kinds are cluster-scoped, which kinds refer to others, etc.
+
+  - Call `KustTarget.makeCustomizedResmap` to create the `ResMap` result.
+
+    This visits everything referenced by the kustomization,
+    performing all generation, transformation and validation.
+
+  - Finish the Run with
+
+    - Optional  `legacy` reordering of objects, overriding
+      a FIFO rule.
+    - Optional addition of _Managed by kustomize_ annotations.
+
+### The `makeCustomizedResmap` function
+
+  This function starts the process of object transformation,
+  as well as accumulation of recursively referenced data.
+
+  - Call `ra := KustTarget.AccumulateTarget`.
+
+    The result, `ra`, is a resource accumulator that contains
+    everything referred to by the current kustomization, now full
+    hydrated.
+
+  - Uniquify names of generated objects by appending content hashes.
+
+    This cannot be done until the objects are complete.
+
+  - Fix all name references (given that names may have changed).
+
+    E.g. if a ConfigMaps was given a generated name, all objects that
+    refer to that ConfigMap must be given it's name.
+
+  - Resolve vars, replacing them with whatever they refer to (a legacy feature).
```

Comment:
How long are we planning to keep in the vars code for?

Reply:
My guess is that it's not so much a matter of time passing as a matter of preparedness to cut major versions. E.g. the vars code needs to go _before_ we can release v1.0 of the API module, but _after_ we release v5 of the CLI.

Reply:
Why does it need to happen after we release v5 of the CLI?

Reply:
Sorry, I meant as part of the v5 release